lib/imagemounter/utils/list_developer_image.js

```diff
@@ -0,0 +1,184 @@
+ * @param {string} name the name of repository
+ * @param {string[]} subFolderList subfolder list in level order
+ * @returns {GithubTreeObject[]?} file list under target folder
+ */
+async function listGithubImageList(user, name, subFolderList = []) {
+    const initialUrl = `https://api.github.com/repos/${user}/${name}/git/trees/master`;
+    const repoUrl = `https://github.com/${user}/${name}/`;
+    const fileList = [];
+    let curUrl = initialUrl;
+    /**
+     * @type {GithubTreeObject[]?}
+     */
+    let ret = null;
+    for (let i = 0; i <= subFolderList.length; i++) {
+        const res = await axios.get(curUrl);
+        /**
+         * @type {GithubTreeResponse}
+         */
+        const body = res.data;
+        const treeList = body?.tree;
+        if (!treeList) {
+            throw new Error(`Failed on looking up ${fileList.join('/')} under ${repoUrl}: ${JSON.stringify(body)}`);
+        }
+        if (i === subFolderList.length) {
+            ret = treeList;
+            break;
+        }
+        const entry = subFolderList[i];
+        const nextItem = _.find(treeList, (item) => item.path === entry);
+        if (!nextItem) {
+            throw new Error(`Unable to find ${entry} under ${fileList.join('/')} in under ${repoUrl}: ${JSON.stringify(treeList)}`);
+        }
+        if (nextItem.type !== 'tree') {
+            throw new Error(`${entry} under ${fileList.join('/')} in ${repoUrl} is expected to be a tree, got ${nextItem.type} instead.`);
+        }
+        fileList.push(entry);
+        curUrl = nextItem.url;
+    }
+    return ret;
+}
+
+/**
+ * Find `DeveloperDiskImage.dmg` recursively under folder and subfolder.
+ * @param {string} entry current folder
+ * @returns {string | undefined} parent folder of `DeveloperDiskImage.dmg` or `undefined` if no such file exists
+ */
+async function findDeveloperImageFromDirectory(entry) {
+    const fileList = await readdir(entry);
+    for (const subEntry of fileList) {
+        const fullPath = joinPath(entry, subEntry);
+        const statResult = await lstat(fullPath);
+        if (subEntry === DEVELOPER_IMAGE_FILE_NAME && statResult.isFile()) {
+            return entry;
+        }
+        if (statResult.isDirectory()) {
+            const subFolderResult = await findDeveloperImageFromDirectory(fullPath);
+            if (subFolderResult) {
+                return subFolderResult;
+            }
+        }
+    }
+    return undefined;
+}
+
+/**
+ * @typedef {Object} ImagePath
+ * @property {string} developerImage
+ * @property {string} developerImageSignature
+ */
+
+/**
+ * Find developer image for certain version. If developer image does not exists,
+ * this will try to find and download developer image, unzip to `~/.appium/iOSSupport/`
+ * @param {string} version full version of iOS device. The first two parts of version
+ * will be preserved when sending the request, e.g. `14.7.1` will be changed to `14.7`
+ * @returns {ImagePath}
```

Comment:
right now we're not typechecking things (but will in the future).  when we do, this will fail, because this function returns a `Promise`.

```suggestion
 * @returns {Promise<ImagePath>}
```